feature:  `add_verbatim_marker()` method to `BraketProgramContext` class for updated `Circuit.from_ir`

sim-eng-ii opened the pull request:
This PR and PR [285](https://github.com/amazon-braket/amazon-braket-default-simulator-python/pull/285) were created to resolve *Issue #972*

*Description of changes:*

In this PR, the following change was made to support the updated `Circuit.from_ir` method:

1. The class `BraketProgramContext(AbstractProgramContext)` was updated to include the method `add_verbatim_marker()`, it invokes:
   - `Instruction(StartVerbatimBox(), target=[])`
   - `Instruction(EndVerbatimBox(), target=[])`

These instructions are called when instances of `VerbatimBoxStart` and `VerbatimBoxEnd` are encountered.  
`add_verbatim_marker()` method overrides the similarly named method from `AbstractProgramContext`, introduced in the `amazon-braket-default-simulator-python` repository in PR [285](https://github.com/amazon-braket/amazon-braket-default-simulator-python/pull/285)

This PR also includes a unit test for the changes to `from_ir`

*Testing done:*

A test function  `test_from_ir_with_verbatim_box()` was added to `test_circuit.py` to verify that `Circuit.from_ir` correctly handles `verbatim_box` sections. 
The test was ran via command:

```bash
pytest test/unit_tests/braket/circuits/test_circuit.py::test_from_ir_with_verbatim_box
```
and 

```bash
pytest test/unit_tests/braket/circuits/test_circuit.py::test_from_ir_with_verbatim_non_verbatim_instr
```

They both pass:

```

test/unit_tests/braket/circuits/test_circuit.py::test_from_ir_with_mixed_verbatim_non_verbatim_instr 
[gw11] [100%] PASSED test/unit_tests/braket/circuits/test_circuit.py::test_from_ir_with_mixed_verbatim_non_verbatim_instr 
====================== 1 passed in 5.89s =====================

test/unit_tests/braket/circuits/test_circuit.py::test_from_ir_with_verbatim_box 
[gw11] [100%] PASSED test/unit_tests/braket/circuits/test_circuit.py::test_from_ir_with_verbatim_box 
====================== 1 passed in 6.11s ======================
```
## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ x] I have read the [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [ x] I used the PR title format described in [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#PR-title-format)
- [ x] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [ x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

sim-eng-ii committed:
Update test_circuit.py

sim-eng-ii committed:
Update test_circuit.py

sim-eng-ii committed:
Merge pull request #1 from robotastray/fix-issue-972

Add unit test `test_from_ir_with_verbatim_box()`for updated `Circuit.from_ir`

robotastray commented:
I am currently completing the changes in my fork of [amazon-braket-default-simulator-python](https://github.com/amazon-braket/amazon-braket-default-simulator-python). I will then open a PR, which should be reviewed alongside the one above.
Thank you.

robotastray commented:
Hello,
I encountered some issues after realising that some of the functions I was importing were causing a circular import error. I am now working on integrating the necessary classes into the `amazon-braket-default-simulator-python` repository. I will continue working on this even after the deadline if I am not able to finish on time. Thank you!

sim-eng-ii committed:
Update BraketProgramCircuit class with add_verbatim_marker method

sim-eng-ii committed:
1st test case

sim-eng-ii committed:
Merge pull request #2 from robotastray/fix-issue-972

Feature: Adding `add_verbatim_marker()` method to `BraketProgramContext` class

sim-eng-ii committed:
test: a`test_from_ir_with_mixed_verbatim_non_verbatim_instr()`

Added an additional test to check for OpenQASM 3.0 with verbatim and non verbatim instructions.

sim-eng-ii committed:
test: `test_from_ir_with_mixed_verbatim_non_verbatim_instr()`

test: added `test_from_ir_with_mixed_verbatim_non_verbatim_instr()`

sim-eng-ii committed:
Merge branch 'main' into main

sim-eng-ii committed:
Fix: updated imports in braket_program_context.py

* update braket_program_context.py with correct imports

* Update braket_program_context.py

rmshaffer committed:
Merge branch 'main' into main

robotastray commented:
hello @rmshaffer I will fix the code format issues

rmshaffer commented:
> The test will not pass until PR [here](https://github.com/amazon-braket/amazon-braket-default-simulator-python/pull/285) is also confirmed and merged to main. The changes in the sdk work only if the [amazon-default-simulator](https://github.com/amazon-braket/amazon-braket-default-simulator-python) is updated

Got it, thanks! If you want to try to pass the CI in the meantime, before that PR is merged, I believe you could temporarily update this line in this PR:
https://github.com/robotastray/amazon-braket-sdk-python/blob/main/tox.ini#L143
to point to your fork, i.e.
```
git+https://github.com/robotastray/amazon-braket-default-simulator-python.git
```

sim-eng-ii committed:
update formatting in `braket_program_context.py` (#5)

* update braket_program_context.py with correct imports

* Update braket_program_context.py

* update to formatting in` braket_program_context.py`

robotastray commented:
> > The test will not pass until PR [here](https://github.com/amazon-braket/amazon-braket-default-simulator-python/pull/285) is also confirmed and merged to main. The changes in the sdk work only if the [amazon-default-simulator](https://github.com/amazon-braket/amazon-braket-default-simulator-python) is updated
> 
> Got it, thanks! If you want to try to pass the CI in the meantime, before that PR is merged, I believe you could temporarily update this line in this PR: https://github.com/robotastray/amazon-braket-sdk-python/blob/main/tox.ini#L143 to point to your fork, i.e.
> 
> ```
> git+https://github.com/robotastray/amazon-braket-default-simulator-python.git
> ```

I have made the changes as you suggested

rmshaffer commented:
@robotastray it's indicating that both branches of the highlighted line are not being hit. To fix that error, you would want tests which cover all branching cases: enter the `if` block, enter the `elif` block, and skip both blocks.

rmshaffer committed:
Merge branch 'main' into main

sim-eng-ii committed:
Fix issue 972  (#6)

* update braket_program_context.py with correct imports

* Update braket_program_context.py

* update to formatting in` braket_program_context.py`

* raise error if unsupported marker is added

* Update test_circuit.py

* fixing formatting and test_circuit.py

ruff check --fix

robotastray commented:
I am currently looking at `test/unit_tests/braket/experimental_capabilities/test_experimental_capabilities.py` to see how my changes have impacted it. I will update once I have more info

robotastray commented:
@rmshaffer the automated test are checking for `test_measureff_ccprx_from_ir`. the function does not exists in `test/unit_tests/braket/experimental_capabilities/iqm/test_iqm_experimental_capabilities.py`

you can see the details in line 6376 on this [page](https://github.com/amazon-braket/amazon-braket-sdk-python/actions/runs/16047937791/job/45299412155?pr=1085)

rmshaffer commented:
> @rmshaffer the automated tests are checking for `test_measureff_ccprx_from_ir`. the function does not exists in `test/unit_tests/braket/experimental_capabilities/iqm/test_iqm_experimental_capabilities.py`
> 
> you can see the details in line 6376 on this [page](https://github.com/amazon-braket/amazon-braket-sdk-python/actions/runs/16047937791/job/45299412155?pr=1085)

@robotastray that test can be found in your fork here: https://github.com/robotastray/amazon-braket-sdk-python/blob/main/test/unit_tests/braket/experimental_capabilities/iqm/test_iqm_experimental_capabilities.py#L190-L206

It's a recently-added test in the upstream repo. I think this failure is expected - your changes are now parsing additional instructions out of the IR. I think you can just update the assertions there so that the test passes again with your changes. For example, replace
```
assert len(circuit_from_ir.instructions) == 3
```
with
```
assert len(circuit_from_ir.instructions) == 5
```
since it now contains the "start" and "end" verbatim box instructions.

sim-eng-ii committed:
Update test_iqm_experimental_capabilities.py (#7)

update `test_iqm_experimental_capabilities.py` to be in line with the incoming changes for PR 972

rmshaffer committed:
Merge branch 'main' into main

sim-eng-ii committed:
Fix test iqm exp for 972 (#8)

* Update test_iqm_experimental_capabilities.py

update `test_iqm_experimental_capabilities.py` to be in line with the incoming changes for PR 972

* update to be test and braket_context

sim-eng-ii committed:
Merge branch 'main' into fixing-conflicts

robotastray commented:
@rmshaffer I updated the code in `test_circuit.py` and `braket_program_context.py` to make sure it is inline with the updates in [#285](https://github.com/amazon-braket/amazon-braket-default-simulator-python/pull/285)

sim-eng-ii committed:
ruff check --fix changes applied

sim-eng-ii reviewed (commented):
braket_program_context with ruff check --fix

**Udate**: you can discard this comment since the changes I have suggested were added in the `ruff check --fix. changes applied`commit

rmshaffer committed:
Merge branch 'main' into fixing-conflicts

rmshaffer committed:
Update tox.ini

rmshaffer committed:
Update tox.ini

rmshaffer committed:
Merge branch 'main' into fixing-conflicts

sim-eng-ii committed:
Merge branch 'main' into fixing-conflicts

speller26 committed:
Move verbatim translation to translations.py

speller26 commented:
Great work!
